src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -720,16 +717,244 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
                                 default(ImmutableArray<string>) :
                                 elementNames.ToImmutableAndFree();
 
-            NamedTypeSymbol tupleTypeOpt = null;
-            var elements = elementTypes.ToImmutableAndFree();
-            var locations = elementLocations.ToImmutableAndFree();
+            return new BoundTupleLiteral(node, boundArguments.ToImmutableAndFree(), elementNamesArray, 
+                                         declaredOrTargetTupleTypeOpt: null, hasErrors: hasErrors);
+        }
+
+        /// <summary>
+        /// Infer natural type for a tuple literal and store relevant information in the node.
+        /// 
+        /// A literal that has a natural type should not add any errors to <paramref name="diagnostics"/>
+        /// during the process.
+        /// </summary>
+        private BoundTupleLiteral AdjustTupleLiteralBasedOnNaturalTypeInference(BoundTupleLiteral tupleLiteral, DiagnosticBag diagnostics, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
```

Comment:
It feels it would be much cheaper to compute natural type of a literal right when we bind it and already have types of elements available.
Deferring this recursive computation makes it much more complicated.